codec/src/types/primitives.rs

```diff
@@ -1,74 +1,120 @@
 //! Implementations of Codec for common types
 
-use crate::{
-    codec::{Codec, Reader, Writer},
-    varint, Error, SizedCodec,
-};
-use bytes::Bytes;
+use crate::{util::at_least, varint, Codec, Error, SizedCodec};
+use bytes::{Buf, BufMut, Bytes};
 use paste::paste;
 
-macro_rules! impl_primitive {
-    ($type:ty, $read_method:ident, $write_method:ident, $size:expr) => {
+// Numeric types implementation
+macro_rules! impl_numeric {
+    ($type:ty, $read_method:ident, $write_method:ident) => {
         impl Codec for $type {
             #[inline]
-            fn write(&self, writer: &mut impl Writer) {
-                writer.$write_method(*self);
+            fn write(&self, buf: &mut impl BufMut) {
+                buf.$write_method(*self);
             }
 
             #[inline]
             fn len_encoded(&self) -> usize {
-                std::mem::size_of::<$type>()
+                Self::LEN_ENCODED
             }
 
             #[inline]
-            fn read(reader: &mut impl Reader) -> Result<Self, Error> {
-                reader.$read_method()
+            fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+                at_least(buf, std::mem::size_of::<$type>())?;
+                Ok(buf.$read_method())
             }
         }
 
         impl SizedCodec for $type {
-            const LEN_ENCODED: usize = $size;
+            const LEN_ENCODED: usize = std::mem::size_of::<$type>();
         }
     };
 }
 
-impl_primitive!(u8, read_u8, write_u8, 1);
-impl_primitive!(u16, read_u16, write_u16, 2);
-impl_primitive!(u32, read_u32, write_u32, 4);
-impl_primitive!(u64, read_u64, write_u64, 8);
-impl_primitive!(u128, read_u128, write_u128, 16);
-impl_primitive!(i8, read_i8, write_i8, 1);
-impl_primitive!(i16, read_i16, write_i16, 2);
-impl_primitive!(i32, read_i32, write_i32, 4);
-impl_primitive!(i64, read_i64, write_i64, 8);
-impl_primitive!(i128, read_i128, write_i128, 16);
-impl_primitive!(f32, read_f32, write_f32, 4);
-impl_primitive!(f64, read_f64, write_f64, 8);
-impl_primitive!(bool, read_bool, write_bool, 1);
+impl_numeric!(u8, get_u8, put_u8);
+impl_numeric!(u16, get_u16, put_u16);
+impl_numeric!(u32, get_u32, put_u32);
+impl_numeric!(u64, get_u64, put_u64);
+impl_numeric!(u128, get_u128, put_u128);
+impl_numeric!(i8, get_i8, put_i8);
+impl_numeric!(i16, get_i16, put_i16);
+impl_numeric!(i32, get_i32, put_i32);
+impl_numeric!(i64, get_i64, put_i64);
+impl_numeric!(i128, get_i128, put_i128);
+impl_numeric!(f32, get_f32, put_f32);
+impl_numeric!(f64, get_f64, put_f64);
+
+// Usize implementation
+impl Codec for usize {
```

Comment:
As noted elsewhere, I strongly recommend removing this.

Reply:
Responded [here](https://github.com/commonwarexyz/monorepo/pull/644#discussion_r2030330897). Resolving this and will use that other thread for discussion